### PR TITLE
Support for Backup Codes

### PR DESCRIPTION
**Pull request by JettChenT**
#676

### CODE REVIEW

**Review by milesmcc — CHANGES_REQUESTED** [review]
Excellent work! See a few small comments here. I might do a small UI pass later, but this looks generally quite good!

File: platform/lib/platform/accounts/user.ex [hunk 1]
```diff
@@ -230,6 +234,29 @@ defmodule Platform.Accounts.User do
     end)
   end
 
+  def update_recovery_codes_changeset(user, attrs) do
+    user
+    |> cast(attrs, [:recovery_codes, :used_recovery_codes])
+  end
+
+  def verify_recovery_code(user, attrs) do
+    code = attrs["current_otp_code"] |> Platform.Utils.parse_recovery_code()
```
[the remark this answers is not in the document]
Comment (JettChenT) [reply platform/lib/platform/accounts/user.ex:243]: Right, updated the corresponding function to handle `nil`

File: platform/lib/platform/accounts/user.ex [hunk 2]
```diff
@@ -230,6 +234,29 @@ defmodule Platform.Accounts.User do
...
+      {:ok,
+       change(user)
+       |> put_change(:recovery_codes, user.recovery_codes -- [code])
+       |> put_change(:used_recovery_codes, user.used_recovery_codes ++ [code])}
+    else
+      {:err, nil}
+    end
+  end
+
+  def check_consume_recovery_changeset(user, attrs) do
```
[the remark this answers is not in the document]
Comment (JettChenT) [reply platform/lib/platform/accounts/user.ex]: Removed.

**Comment by JettChenT** [comment]
Note:
- Not sure if recovery codes can be used to access `settings/mfa` page, generate new recovery codes, and disable MFA. (currently it can)
- Regarding https://github.com/atlosdotorg/atlos/pull/828/commits/27207f66d7be716b39a22466fded6d7ae1b99fe2 , it seems pretty hard to get `phx-disable-with` working with buttons with icons, as it directly replaces the `innerText` of the element.

File: platform/lib/platform_web/live/settings_live/mfa_live.ex [hunk 3]
```diff
@@ -58,6 +59,20 @@ defmodule PlatformWeb.SettingsLive.MFALive do
     end
   end
 
+  def handle_event("access_otp", %{"access_otp" => query}, socket) do
+    changeset = Accounts.confirm_user_mfa(socket.assigns.current_user, query)
+
+    if changeset.valid? do
+      {:noreply, socket |> assign(:otp_passed, true)}
```
Comment (milesmcc) [inline platform/lib/platform_web/live/settings_live/mfa_live.ex:66]: Currently you can use a backup code to pass this check but the backup code is not immediately marked as invalid in the interface.

**Review by milesmcc — CHANGES_REQUESTED** [review]
Excellent work! Nearly there. Pushed a few small design/formatting changes myself.

File: platform/lib/platform_web/live/settings_live/mfa_live.ex (hunk 3 above)
Comment (JettChenT) [reply platform/lib/platform_web/live/settings_live/mfa_live.ex]: Fixed by applying changes and updating `current_user`

File: platform/lib/platform_web/live/settings_live/backup_codes_component.ex [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    <div>
+      <.card>
+        <:header>
+          <h3 class="sec-head">Backup Codes</h3>
+          <p class="sec-subhead">
+            Generate backup codes to use in case you lose access to your authenticator app.
+          </p>
+        </:header>
+        <p>
+          <%= if length(@current_user.recovery_codes)>0 || length(@current_user.used_recovery_codes)>0 do %>
```
[the remark this answers is not in the document]
Comment (JettChenT) [reply platform/lib/platform_web/live/settings_live/backup_codes_component.ex]: Indeed! Applied

File: platform/lib/platform_web/controllers/media/export_controller.ex [hunk 5]
```diff
@@ -215,4 +216,31 @@ defmodule PlatformWeb.ExportController do
 
     conn
   end
+
+  def create_backup_codes_export(conn, _params) do
```
[the remark this answers is not in the document]
Comment (JettChenT) [reply platform/lib/platform_web/controllers/media/export_controller.ex]: Implemented the signed Token approach.

File: platform/lib/platform_web/controllers/accounts/user_session_controller.ex [hunk 6]
```diff
@@ -1,4 +1,5 @@
 defmodule PlatformWeb.UserSessionController do
+  require Logger
```
[the remark this answers is not in the document]
Comment (JettChenT) [reply platform/lib/platform_web/controllers/accounts/user_session_controller.ex]: Deleted

File: platform/lib/platform/utils.ex [hunk 7]
```diff
@@ -255,6 +255,34 @@ defmodule Platform.Utils do
...
+    |> Enum.join(" ")
+  end
+
+  def parse_recovery_code(code) do
+    if code == nil do
+      nil
+    else
+      code
+      |> String.split(" ", trim: true)
+      |> Enum.join("")
```
[the remark this answers is not in the document]
Comment (JettChenT) [reply platform/lib/platform/utils.ex]: Yes! Applied

**Review by milesmcc — APPROVED** [review]
Well done!

**Pull request merged**